### PR TITLE
matplotlib 3.9.2

### PR DESCRIPTION
**Pull request by skupr-anaconda**
**Destination channel:** main

### Links

- [PKG-5122](https://anaconda.atlassian.net/browse/PKG-5122) 
- [Upstream repository](https://github.com/matplotlib/matplotlib/tree/v3.9.2)
- Requirements: 
  - https://github.com/matplotlib/matplotlib/blob/v3.9.2/pyproject.toml
  - https://matplotlib.org/stable/install/dependencies.html#build-dependencies

### Explanation of changes:

- Rewrite build scripts (taken from the conda-forge recipe) because `matplotlib` uses `python-meson` as a build backend now. The only difference is to add `--no-deps --no-build-isolation --no-index` to `pip install`. It will prevent downloading dependencies from `pip`.
  - Add `export AR=$GCC_AR` because it's important on Linux, see https://github.com/matplotlib/matplotlib/issues/28260#issuecomment-2146243663
- Remove a local cbc.yaml as obsolete
- Remove obsolete `000_use_github_url_qhull.patch ` because `setupext.py` doesn't exist already
- Update `host` and `run` dependencies
- Remove the `mpl_sample_data` subpackage  and `install-data.(sh|bat)` scripts because the data files are available in the `matplotlib-base` subpackage by defaults
  - Add `matplotlib.cbook` to `test/imports` (moved from the deleted `mpl_sample_data` subpackage)
  - Move the command `python -c "import matplotlib; matplotlib.cbook.get_sample_data('msft.csv', asfileobj=False)"` from  `mpl_sample_data` to `matplotlib-base`

[PKG-5122]: https://anaconda.atlassian.net/browse/PKG-5122?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by skupr-anaconda** [comment]
An error on linux-64:
`meson-python: error: Could not find meson version 0.63.3 or newer, found . `

**Comment by skupr-anaconda** [comment]
An error:
```
 conda_build.exceptions.OverLinkingError: overlinking check failed

    ['  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libwebpmux-376a476e.so.3.1.0): $RPATH/libwebp-7ac63b4c.so.7.1.9 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libwebp-7ac63b4c.so.7.1.9): $RPATH/libsharpyuv-f605eae1.so.0.1.0 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libfreetype-d604525d.so.6.20.1): $RPATH/libpng16-cdae5f13.so.16.43.0 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', "  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libfreetype-d604525d.so.6.20.1): .. but ['defaults/linux-aarch64::zlib==1.2.13=h998d150_1'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", '  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libfreetype-d604525d.so.6.20.1): $RPATH/libbrotlidec-16e97476.so.1.1.0 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libharfbuzz-c838b13c.so.0.60850.0): $RPATH/libfreetype-d604525d.so.6.20.1 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libwebpdemux-5e1da4e9.so.2.0.15): $RPATH/libwebp-7ac63b4c.so.7.1.9 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', "  ERROR (matplotlib-base,lib/python3.12/site-packages/PIL/_imaging.cpython-312-aarch64-linux-gnu.so): .. but ['defaults/linux-aarch64::zlib==1.2.13=h998d150_1'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", '  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libtiff-bdf4bcf3.so.6.0.2): $RPATH/liblzma-0c11df12.so.5.4.5 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libtiff-bdf4bcf3.so.6.0.2): $RPATH/libjpeg-622c4989.so.62.4.0 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', "  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libtiff-bdf4bcf3.so.6.0.2): .. but ['defaults/linux-aarch64::zlib==1.2.13=h998d150_1'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", '  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libxcb-c830e669.so.1.1.0): $RPATH/libXau-21870672.so.6.0.0 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', "  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libpng16-cdae5f13.so.16.43.0): .. but ['defaults/linux-aarch64::zlib==1.2.13=h998d150_1'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)", '  ERROR (matplotlib-base,lib/python3.12/site-packages/pillow.libs/libbrotlidec-16e97476.so.1.1.0): $RPATH/libbrotlicommon-526214ec.so.1.1.0 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?']
```

**Comment by skupr-anaconda** [comment]
An error:
```
/croot/matplotlib-suite_1724924833987/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/include/pybind11/detail/common.h:274:10: fatal error: Python.h: No such file or directory

  274 | #include <Python.h>
```

**Review by cbouss — APPROVED** [review]

File: recipe/meta.yaml [hunk 1]
```diff
@@ -28,34 +28,33 @@ outputs:
...
         - python
         - pip
-        - certifi >=2020.06.20
-        - freetype 2.10
-        - numpy {{ numpy }}
-        - setuptools 
+        - freetype {{ freetype }}
+        - numpy 1.23         # [py<311]
+        - numpy {{ numpy }}  # [py>=311]
         - setuptools_scm >=7
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml:40]: This should not be needed anymore.

Comment (skupr-anaconda) [reply recipe/meta.yaml:40]: `setuptools_scm` is still in the list https://github.com/matplotlib/matplotlib/blob/v3.9.2/pyproject.toml#L77 and meson checks if it's avaialble https://github.com/matplotlib/matplotlib/blob/v3.9.2/meson.build#L4

Comment (skupr-anaconda) [reply recipe/meta.yaml:40]: numpy is also there https://github.com/matplotlib/matplotlib/blob/v3.9.2/pyproject.toml#L81-L89. Could you clarify what exactly not needed?

Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:40]: Oh, NVM then.

**Review by Jrice1317 — APPROVED** [review]
LGTM

**Pull request merged**